Add new source for Viber Desktop local sqlite db

ankostis opened the pull request:
# Add new source for Viber Desktop local sqlite db

Viber Desktop (at least in linux that i checked) stores its messages in a local SQLite file 
`~/.ViberPC/{user-phone-number}/viber.db` and it contains everything needed to devise Visits stream,
as shown in the the sample below:
![image](https://user-images.githubusercontent.com/501585/108750207-6b686c00-7549-11eb-9f66-e5d860cfa5b8.png)

The indexer collects:

- chat-name
- sender
- chat text
- title extracted from any previewed url
- chat-tags
- any URLs on the chat-text

The only one missing is a working `Locator.href`, which i couldn't think of anything other than:

    sqlite:///home/username/.ViberPC/123456789/viber.db?immutable=1#!Messages.EventId=3445` 

It is configured with:

```python
SOURCES = [
    Source(viber_desktop.index, "~/.ViberPC/your-phone-number/viber.db"),
]
```

* DOCs.have been updated;
* Sample config has been updated.

ankostis committed:
FEAT: +Viber source from local sqlite db

ankostis committed:
refact(viber) mv viber source to .._desktop

which is the correct name for it.

ankostis committed:
enh(viber) resolve homedir on db-path

ankostis committed:
doc(viber) sample-cfg & src-section

ankostis committed:
upd(viber) rename back to plain `viber`  ...

bc cannot control src-name otherwise.

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+        try:
+            yield from _handle_row(row)
+        except Exception as ex:
+            logger.warning(
```
Replying to your comment [here](https://github.com/karlicoss/promnesia/commit/4363b5797b30819b89a4e4908ea27b009b598112#commitcomment-47433572) 

Basically the options people usually use are:

1. throw immediately (as you suggested in the comment)
  This is problematic because then on any unexpected situation (like 1/10000 messages parsed incorrectly), the program will break. Considering we have to deal with closed protocols, formats change without warning, and it's annoying when your software breaks until you fix it, so we need to be more flexible and defensive here. In addition, imagine if one data source breaks the whole program -- sometimes it's justified (e.g. if you were working with important medical data or something like that), but not in case of this tool.

2. log and continue (as you're doing here)
   It works nice in terms of not breaking immediately. However then there is another danger -- imagine that for example Viber changes format for 'time' column to store the isoformatted string (instead of epoch), for whatever reason. Then promnesia indexer will carry on working (because of exception catching) -- it's good. However, you won't notice the error unless you regularly reviewing the logs, or maybe when you know for sure that Promnesia should fire but it doesn't.

So in Promnesia (and related projects that process data, like HPI) I'm doing something intermediate.

- catch the exceptions (where it makes sense, e.g. parsing)
- possibly log it if necessary
- but then `yield` it to the parent function, as a normal value (there is even an alias for [`Result`](https://github.com/karlicoss/promnesia/blob/9b7520feea9872e29520d466aa302b76ea068c52/src/promnesia/common.py#L128-L130) )
 
This allows the top level handler to exit with code 1 (so you get a cron email/systemd job failure etc and notice the error), while still doing what's the program supposed to do (process as much as it can and insert in the database).

As more examples why it's useful in Promnesia, errors are also inserted in the database:
- this makes it easier to test in some cases (instead of intercepting logs, just check the database)
- later it will allow to overview the errors form the browser which would be much more convenient for most users than messing with systemd logs or whatnote

I write more about it [here](https://beepb00p.xyz/mypy-error-handling.html), you might enjoy it!

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+        if not urls:
+            return
+        dt = from_epoch(row["time"] // 1000)  # timestamps are stored x100 this db
+        mid: str = unwrap(row["mid"])
```
Oh! my bad, for `unwrap` in telegram module I intented to do `notnone` instead, [like in HPI](https://github.com/karlicoss/HPI/blob/ad924ebca84b0846c98401b10e61488c45cf1e9c/my/core/error.py#L18-L20). The reason was to make sure null rows aren't converted to `None` in python and avoid downstream errors.

`Exception` can never be in the sqlite row, so you can either remove `unwrap` from this module, or use `notnone` instead, up to you.

ankostis replied:
I was having this question :-)

ankostis replied:
Resolved in a8c6dd0.

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+    # TODO context manager?
+    db = dataset_readonly(path)  # TODO could check is_file inside
+
+    query_str = textwrap.dedent(
```
nice, didn't know about `dedent` (and the whole textwrap module!)

ankostis replied:
One day i may submit a mass-update PR for all sources needing `dedent` affection :-)

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+from ..common import (Loc, PathIsh, Results, Visit, echain, extract_urls,
+                      from_epoch, get_logger)
+
+# TODO potentially, belongs to my. package
```
Ideally this module (i.e. the query/parsing part) would be part of HPI (so viber data can be used by other apps).
The only reason it's not is that the promnesia module existed before I started HPI project :)
But totally ok to keep here for now, can always refactor later if needed.

ankostis replied:
(read it before my own suggestion on that, [below](https://github.com/karlicoss/promnesia/pull/204#pullrequestreview-597263776)).
Yes i agree.
Will transplant part of this PR in HPI.

I will wait for your review to finish here, before spreading stuff in 2 projects.

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+    return dataset.connect("sqlite:///", engine_kwargs={"creator": creator})
+
+
+def index(database: PathIsh) -> Results:
```
One nice thing to do could be to make it optional , and it it's None, then do glob over `"~/.ViberPC/*/viber.db"`, then the user could just specify `Source(viber_desktop.index)` in the config without extra arguments.
But up to you.

ankostis replied:
Nice idea!
[edit] actually `SOURCES=[..., viber]` would suffice.

ankostis replied:
Implemented in 2d5c3c4.

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+            # https://www.reddit.com/r/Telegram/comments/6ufwi3/link_to_a_specific_message_in_a_channel_possible/
+            # hmm, only seems to work on mobile app, but better than nothing...
+            yield Visit(
+                url=unquote(u),
```
are they quoted in the viber database? Might be unnecessary (and if they are do you mind adding a comment?)

ankostis replied:
No they are not, removed it.
Resolved in f4553ed.

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+                context=text,
+                locator=Loc.make(
+                    title=f"chat({mid}) from {sender}@{chatname}",
+                    href=f"sqlite://{database}?immutable=1#!Messages.EventId={mid}",
```
interesting! does it actually work, i.e. if you press on the link it can jump to this message in sqlitebrowser? (or what are you using?)

ankostis replied:
Ofcourse not :-)
 
But i needed something to contain internal information to help the user locate the message, besides the chat-name & the date, already visible in the sidebar.  And i wanted to be clear that it requires _programmatic-access_, hence the `sqlite` url-schema, which can be used verbatim as a connection URI.
I think the fragment with the `table.column=<mid>` criterio would not matter (but neither work), but haven't checked it.

Is there some other idea (like telegram forwarding- url service)?

ankostis replied:
I managed to open the Viber DB in `sqlitebrowser` in linux, 
by ifirst making a "hard" link to the db-file, replacing the suffix from `.db --> .sqlite`,
an then simply referring to it with `file:///...sqlite` uri.

Would that be ok (for `promnesia index` to make a hard-link in huser's viber-config directory??

ankostis replied:
I implemented the `file:://...sqlite` locator-URIs in df948c3.

Unfortunately, the links do not work, due to browser's security preventing opening local file-links, i guess,
because when i copy-past the link in another tab, it does offer me to open it with `sqlitebrowser`.

karlicoss replied:
> I think the fragment with the table.column=<mid> criterio would not matter

Yeah fair enough. Just asked in case it was some hidden feature, it could be convenient :) 

I think the closest possible thing that sqlitebrowser has is `--sql` argument (although I couldn't get it to work), but it wouldn't work file `file://` handler unfortunately.

> Is there some other idea (like telegram forwarding- url service)?

Quick search gives something: https://stackoverflow.com/a/41324451/706389, but it's probably unlikely to work on desktop unfortunately :( Such things are always better on mobile.

> Would that be ok (for promnesia index to make a hard-link in huser's viber-config directory??

Hmm. Hardlinks are pretty extreme and opaque IMO -- the user might damage the original database without knowing it. But I see the problem with symlinks, they seem to be resolved before opening and it then tries to open '.db' file.
Maybe comment the hardlink logic for now? I'm worried that someone trashes their database; maybe we can come up with a more robust idea later? 
Also see the comment https://github.com/karlicoss/promnesia/pull/204#issuecomment-785374009 how to handle it on user site without hardlinking (a little tedious, but you only need to do it once, and maybe later we can integrate these instructions in promnesia, similar to [open-with-editor(https://github.com/karlicoss/open-in-editor) ?)

ankostis replied:
Great research.
Will drop the hard-link part.  Closing this review.

karlicoss commented on src/promnesia/sources/viber_desktop.py:
```diff
@@ -0,0 +1,164 @@
+
+        for u in urls:
+            # https://www.reddit.com/r/Telegram/comments/6ufwi3/link_to_a_specific_message_in_a_channel_possible/
+            # hmm, only seems to work on mobile app, but better than nothing...
```
can remove this comment I guess

ankostis replied:
Resolved in 4c2c5d6.

karlicoss commented on doc/SOURCES.org:
```diff
@@ -136,6 +136,11 @@ for importer, name, ispkg in sorted(pkgutil.walk_packages(
 
+- [[file:../src/promnesia/sources/viber_desktop.py][promnesia.sources.viber_desktop]]
+
+  Uses the local Viber Desktop SQLite databse in your home directory, 
```
typo in 'database'

ankostis replied:
Fixed in 2d5c3c4.

karlicoss reviewed (changes requested):
Very nice, thanks so much, especially for figuring out the query! 
Looks good, there are some inor comments (and some optional, so up to you if you want to fix them, otherwise I can clean up sometime later).

I don't have Viber so can't test it (to ensure it doesn't break later). Do you know by any chance, if there some public viber database somewhere on github, so it would be possible to run some basic tests against it? Otherwise not a big problem, hopefully the most potentially fragile part (query) won't need to change at all for a while.

ankostis committed:
doc(viber): minor sql & python comments

ankostis committed:
style(viber) black-format

ankostis committed:
fix(viber)  urls in sqlite are not quoted

ankostis committed:
fix(viber) don'unwrap not needed from sqlite, ...

enhance sanity checks on empty viber fields,
ass comented by @karlicoss in karlicoss/promnesia#204.

ankostis reviewed (commented):
>  Do you know by any chance, if there some public viber database somewhere on github

No.
I could empty mine quasi-completely, to keep just some for running the query, and upload it here or the creation sql-scripts.
Even better if i could upload the DDL+insert statements it in the sources, to create test-case in the sources.

Besides this requiring some time-investment from my side, it leads to another question:
maybe se shpould move the query part of the code in HPI, no?

ankostis committed:
fix(viber) locator-uri includes resolved sqlite-path

ankostis committed:
refact(viber) msgs-query in own func,

... move index() closures to the top of  the function.

ankostis committed:
fix(viber) open db in context-manager (old TODO)

ankostis committed:
ENH(viber) hard-link db-->.sqlite for locator-urls

ankostis committed:
doc: sample cfg use module, not module.index

ankostis committed:
refact(viber) uneeded closure

ankostis committed:
refact(viber) un-closureify handle_row()

ankostis committed:
refact(viber) privatize open_db(); doc

ankostis committed:
enh(viber) log db hard-link creation

ankostis committed:
refact(viber) org imports, drop dead code

ankostis committed:
ENH(viber) support multiple GLOBbed DB-paths,
...and use a sensible default.

ankostis committed:
drop(viber) hard-link hack from .vb --> .sqlite,

prefer @karlicoss's teaching firefox to recognize `.db` mime-type:
https://github.com/karlicoss/promnesia/pull/204#issuecomment-785374009
